wgpu-native/src/id.rs

```diff
@@ -0,0 +1,111 @@
+use crate::{Epoch, Id, Index, TypedId};
+use std::marker::PhantomData;
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+#[repr(transparent)]
+#[derive(Clone, Copy, Debug, Hash, PartialEq)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct GenericId<T> {
+    id: Id,
+    _phantom: std::marker::PhantomData<T>,
+}
+
+impl<T> GenericId<T> {
+    fn raw(&self) -> Id {
+        self.id
+    }
+}
+
+impl <T> TypedId for GenericId<T> {
+    fn new(index: Index, epoch: Epoch) -> Self {
+        Self {
+            id: Id(index, epoch),
+            _phantom: PhantomData,
+        }
+    }
+
+    fn index(&self) -> Index {
+        (self.raw()).0
+    }
+
+    fn epoch(&self) -> Epoch {
+        (self.raw()).1
+    }
+}
+
+type HandleId<T> = GenericId<PhantomData<T>>;
```

Comment:
it appears that we use `PhantomData<PhantomData<T>>` in the end? (one here, one inside `GenericId`)
it looks like `HandleId` isn't actually needed

Reply:
Right, I'm trying to avoid having two levels of PhantomData but I'm hitting an issue with derives (maybe due to https://github.com/rust-lang/rust/issues/26925). I need to keep experimenting with this a bit